## QA-01. Annotation QA(라벨링 품질검수) 목적

### 목적

본 단계에서는 Pilot Object Labeling (파일럿 객체 라벨링) 결과의 **정확성, 완전성, 일관성**을 검증한다.

단순한 데이터 구조 오류뿐만 아니라 Bounding Box(바운딩 박스), Class Label(클래스 라벨), 
객체 속성 및 라벨링 기준의 일관성을 점검하여 향후 데이터 분석 및 모델 학습에 활용할 수 있는 품질 수준인지 평가한다. 

### 주요 검수 범위

1. **Schema Validation(스키마 검증)**
   - 필수 컴럼과 데이터 구조가 올바르게 구성되어 있는지 확인
   - Missing Value(결측값) 및 잘못된 데이터 형식 확인

2. **Bounding Box Validation(바운딩 박스 검증)**
   - `x_min < x_max`, `y_min < y_max` 조건 확인
   - 바운딩 박스 좌표가 이미지 범위를 벗어나지 않는지 확인

3. **Class Label Validation(클래스 라벨 검증)**
   - 사전에 정의된 클래스만 사용되었는지 확인
   - 오탈자 및 일관되지 않은 클래스명 탐지

4. **Object ID Validation(객체 ID 검증)**
   - 중복된 Object ID(객체 ID) 확인
   - 객체 식별 규칙이 일관되게 적용되었는지 검증

5. **Attribute Validation(속성 검증)**
   - Occlusion(가림), Truncation(잘림), Small/Distant(소형·원거리), Ambiguity(모호성) 값 검증
   - 정의되지 않는 값 또는 비정상적인 속성 조합 확인 

6. **Annotation Consistency(라벨링 일관성)**
   - 유사한 객체에 동일한 라벨링 기준이 적용되었는지 확인 
   - 추가 검토가 필요한 Edge Case(경계 사례) 식별

7. **Visual QA(시각적 품질검수)**
   - 이미지에 Bounding Box(바운딩 박스)와 Class Label(클래스 라벨)을 표시 
   - 객체 포함 범위, 박스 위치 및 클래스 정확성을 육안으로 검토 

### 기대 결과 

본 QA(품질검수)를 통해 다음 결과를 도출한다. 

- 전체 라벨링 수 및 유효/오류 라벨링 수
- Bounding Box(바운딩 박스) 유효성 검증 결과
- Class Label(클래스 라벨) 및 Attribute(속성) 검증 결과
- 잠재적인 라벨링 오류 및 Edge Case(경계 사례)
- QA Pass Rate(품질검수 통과율)
- Annotation Guideline(라벨링 가이드라인) 개선 사항  


## QA-02. Data Loading & Initial Check(데이터 불러오기 및 초기 점검)

### 목적

Annotation QA(라벨링 품질검수)에 사용할 Pilot Labeling Data(파일럿 라벨링 데이터)를 
불러오고 데이터의 기본 구조와 레코드 수를 확인한다. 

Data Understanding(데이터 이해) 단계에서 확인한 결과와 비교하여, QA 과정에서 올바른
데이터 파일이 사용되고 있는지 검즘한다. 

### 주요 확인 항목 

- Dataset Shape(데이터셋 크기)
- 전체 Object Annotation(객체 라벨링) 수
- Unique Image(고유 이미지) 수
- Column(컬럼) 구성
- 데이터 Sample(샘플) 확인

In [1]:
import pandas as pd
from pathlib import Path

# 현재 작업 경로 확인
print("Current Working Directory:")
print(Path.cwd())

Current Working Directory:
c:\Users\user\OneDrive\Documents\data-labeling-portfolio\03_autonomous_driving\notebooks


In [2]:
# 상위 프로젝트 폴더의 파일 구조 확인
project_root = Path.cwd().parent

print("Project Root:")
print(project_root)

print("\nFiles and Folders:")
for item in sorted(project_root.iterdir()):
    print("-", item.name)

Project Root:
c:\Users\user\OneDrive\Documents\data-labeling-portfolio\03_autonomous_driving

Files and Folders:
- annotations
- data
- docs
- images
- notebooks
- README.md


In [3]:
# Annotations 폴더 내부 파일 확인
annotations_dir = project_root / "annotations"

print("Annotations Directory:")
print(annotations_dir)

print("\nAnnotation Files:")
for item in sorted(annotations_dir.iterdir()):
    print("-", item.name)

Annotations Directory:
c:\Users\user\OneDrive\Documents\data-labeling-portfolio\03_autonomous_driving\annotations

Annotation Files:
- pilot


In [4]:
# Pilot 폴더 내부 파일 확인
pilot_dir = annotations_dir / "pilot"

print("Pilot Annotation Directory:")
print(pilot_dir)

print("\nPilot Annotation Files:")
for item in sorted(pilot_dir.iterdir()):
    print("-", item.name)

Pilot Annotation Directory:
c:\Users\user\OneDrive\Documents\data-labeling-portfolio\03_autonomous_driving\annotations\pilot

Pilot Annotation Files:
- cvat_pilot_annotations.zip
- main
- manual_annotations_73.csv


In [5]:
# Pilot Labeling Data(파일럿 라벨링 데이터) 불러오기 
annotation_file = pilot_dir / "manual_annotations_73.csv"

df = pd.read_csv(annotation_file)

print("=== Pilot Annotation Data Loaded ===")
print(f"File: {annotation_file.name}")
print(f"Dataset Shape: {df.shape}")
print(f"Total Annotations: {len(df)}")
print(f"Unique Images: {df['image_name'].nunique()}")

print("\n=== Columns ===")
print(df.columns.tolist())

print("\n=== Data Sample ===")
display(df.head())

=== Pilot Annotation Data Loaded ===
File: manual_annotations_73.csv
Dataset Shape: (73, 12)
Total Annotations: 73
Unique Images: 8

=== Columns ===
['image_name', 'object_id', 'class_label', 'x_min', 'y_min', 'x_max', 'y_max', 'occlusion', 'truncation', 'small_distant', 'ambiguity', 'notes']

=== Data Sample ===


,image_name,object_id,class_label,x_min,y_min,x_max,y_max,occlusion,truncation,small_distant,ambiguity,notes
0,image_01.jpg,C01_OBJ_001,Traffic Cone,719.70,495.10,736.20,562.38,NaN,NaN,False,NaN,NaN
1,image_01.jpg,C01_OBJ_002,Barrier,2.03,463.32,255.79,582.82,Partial,Partial,False,NaN,NaN
2,image_01.jpg,C01_OBJ_003,Truck,280.60,456.43,526.20,573.40,Partial,NaN,False,NaN,NaN
3,image_01.jpg,C01_OBJ_004,Traffic Cone,596.90,492.10,630.00,555.50,Partial,NaN,False,NaN,NaN
4,image_01.jpg,C01_OBJ_005,Traffic Cone,681.69,493.17,698.17,564.33,NaN,NaN,False,NaN,NaN


## QA-03. Schema Validation(스키마 검증)

### 목적

Pilot Labeling Data (파일럿 라벨링 데이터)의 구조가 사전에 정의한 
Annotation Schema(라벨링 스키마)를 충족하는지 검증한다. 

필수 컬럼의 존재 여부, Data Type(데이터 타입), Missing Value(결측값), 
Duplicate Record(중복 레코드)를 확인하여 구조적인 데이터 품질 문제를 식별한다.

### 주요 검증 항목

- Required Column(필수 컬럼) 존재 여부
- Column Count(컬럼 수) 확인
- Data Type(데이터 타입) 확인
- Missing Value(결측값) 확인 
- Duplicate Record(중복 레코드) 확인 

In [6]:
# Required Column(필수 컬럼) 검증

required_columns = [
    "image_name",
    "object_id", 
    "class_label",
    "x_min",
    "y_min",
    "x_max",
    "y_max",
    "occlusion",
    "truncation",
    "small_distant",
    "ambiguity",
    "notes"
]

missing_columns = [
    col for col in required_columns
    if col not in df.columns
]

extra_columns = [
    col for col in df.columns
    if col not in required_columns
]

print("=== Required Column Validation ===")
print(f"Expected Columns: {len(required_columns)}")
print(f"Actual Columns: {len(df.columns)}")
print(f"Extra Columns: {extra_columns}")

if not missing_columns and not extra_columns:
    print("\nResult: PASS - Schema columns match the expected structure.")
else: 
    print("\nResult: REVIEW - Schema column mismatch detected.")

=== Required Column Validation ===
Expected Columns: 12
Actual Columns: 12
Extra Columns: []

Result: PASS - Schema columns match the expected structure.


In [7]:
# Data Type(데이터 타입) 확인

print("=== Data Type Validation ===")
print(df.dtypes)

print("\n=== Data Type Groups ===")
print(f"Numeric Columns: {df.select_dtypes(include='number').columns.tolist()}")
print(f"Object Columns: {df.select_dtypes(include='object').columns.tolist()}")
print(f"Boolean Columns: {df.select_dtypes(include='bool').columns.tolist()}")

=== Data Type Validation ===
image_name           str
object_id            str
class_label          str
x_min            float64
y_min            float64
x_max            float64
y_max            float64
occlusion            str
truncation           str
small_distant       bool
ambiguity            str
notes            float64
dtype: object

=== Data Type Groups ===
Numeric Columns: ['x_min', 'y_min', 'x_max', 'y_max', 'notes']
Object Columns: ['image_name', 'object_id', 'class_label', 'occlusion', 'truncation', 'ambiguity']
Boolean Columns: ['small_distant']


C:\Users\Public\Documents\ESTsoft\CreatorTemp\ipykernel_6280\205067001.py:8: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print(f"Object Columns: {df.select_dtypes(include='object').columns.tolist()}")


In [8]:
# Data Type(데이터 타입) 확인

print("=== Data Type Validation ===")
print(df.dtypes)

print("\n=== Data Type Groups ===")
print(f"Numeric Columns: {df.select_dtypes(include='number').columns.tolist()}")
print(f"String Columns: {df.select_dtypes(include='str').columns.tolist()}")
print(f"Boolean Columns: {df.select_dtypes(include='bool').columns.tolist()}")

=== Data Type Validation ===
image_name           str
object_id            str
class_label          str
x_min            float64
y_min            float64
x_max            float64
y_max            float64
occlusion            str
truncation           str
small_distant       bool
ambiguity            str
notes            float64
dtype: object

=== Data Type Groups ===
Numeric Columns: ['x_min', 'y_min', 'x_max', 'y_max', 'notes']
String Columns: ['image_name', 'object_id', 'class_label', 'occlusion', 'truncation', 'ambiguity']
Boolean Columns: ['small_distant']


In [9]:
# Missing Value(결측값) 검증

missing_count = df.isna().sum()
missing_rate = (df.isna().mean() * 100).round(2)

missing_summary = pd.DataFrame({
    "missing_count": missing_count,
    "missing_rate_pct": missing_rate
})

print("=== Missing Value Validation ===")
display(missing_summary)

print(f"\nTotal Missing Values: {df.isna().sum().sum()}")

=== Missing Value Validation ===


,missing_count,missing_rate_pct
image_name,0,0.00
object_id,0,0.00
class_label,0,0.00
x_min,0,0.00
y_min,0,0.00
x_max,0,0.00
y_max,0,0.00
occlusion,32,43.84
truncation,68,93.15
small_distant,0,0.00



Total Missing Values: 226


In [10]:
# Attribute(속성) 컬럼의 실제 값 분포 확인

attribute_columns = [
    "occlusion",
    "truncation",
    "small_distant",
    "ambiguity",
    "notes"
]

print("=== Attribute Value Distribution ===")

for col in attribute_columns:
    print(f"\n[{col}]")
    print(df[col].value_counts(dropna=False))

=== Attribute Value Distribution ===

[occlusion]
occlusion
NaN        32
Partial    31
Heavy      10
Name: count, dtype: int64

[truncation]
truncation
NaN        68
Partial     4
Heavy       1
Name: count, dtype: int64

[small_distant]
small_distant
True     41
False    32
Name: count, dtype: int64

[ambiguity]
ambiguity
NaN         53
Boundary    12
Class        8
Name: count, dtype: int64

[notes]
notes
NaN    73
Name: count, dtype: int64


In [11]:
# Semantic Missing Value(의미적 결측값) 검증

# 빈 값이 허용되는 Optional / None-state 컬럼
allowed_empty_columns = [
    "occlusion",
    "truncation",
    "ambiguity",
    "notes"
]

# 반드시 값이 존재해야 하는 Required(필수) 컬럼
required_value_columns = [
    "image_name",
    "object_id",
    "class_label",
    "x_min",
    "y_min",
    "x_max",
    "y_max",
    "small_distant"
]

required_missing = df[required_value_columns].isna().sum()
total_required_missing = required_missing.sum()

print("=== Semantic Missing Value Validation ===")

print("\n[Required Fields]")
print(required_missing)

print("\n[Optional / None-state Fields]")
print(df[allowed_empty_columns].isna().sum())

print(f"\nActual Missing Values in Required Fields: {total_required_missing}")

if total_required_missing == 0:
    print("Result: PASS - No missing values detected in required fields.")
else:
    print("Result: REVIEW - Missing values detected in required fields.")

=== Semantic Missing Value Validation ===

[Required Fields]
image_name       0
object_id        0
class_label      0
x_min            0
y_min            0
x_max            0
y_max            0
small_distant    0
dtype: int64

[Optional / None-state Fields]
occlusion     32
truncation    68
ambiguity     53
notes         73
dtype: int64

Actual Missing Values in Required Fields: 0
Result: PASS - No missing values detected in required fields.


In [12]:
# Duplicate Record(중복 레코드) 검증

# 전체 행 기준 중복
duplicate_rows = df.duplicated().sum()

# Object ID(객체 ID) 기준 중복
duplicate_object_ids = df["object_id"].duplicated().sum()

print("=== Duplicate Record Validation ===")
print(f"Duplicate Rows: {duplicate_rows}")
print(f"Duplicate Object IDs: {duplicate_object_ids}")

if duplicate_rows == 0 and duplicate_object_ids == 0:
    print("\nResult: PASS - No duplicate records or object IDs detected.")
else:
    print("\nResult: REVIEW - Duplicate records or object IDs detected.")

=== Duplicate Record Validation ===
Duplicate Rows: 0
Duplicate Object IDs: 0

Result: PASS - No duplicate records or object IDs detected.


### QA-03 Summary(요약)

Schema Validation(스키마 검증) 결과, Pilot Labeling Data(파일럿 라벨링 데이터)는 
정의된 데이터 구조와 필수 품질 기준을 충족하였다. 

- **Required Columns(필수 컬럼):** 12개의 모두 존재하며 누락 및 추가 컬럼 없음 → **PASS** 
- **Data Type(데이터 타입):** Bounding Box(바운딩 박스) 좌표는 수치형, 주요 라벨 및 속성은 문자열/줄리언 형태로 정상 구성
- **Required Field Missing Values(필수 필드 결측값):** 0건 → **PASS**
- **Duplicate Object IDs(중복 객체 ID):** 0건 → **PASS**

### QA Finding(품질검수 발견사항)

CSV 저장 과정에서 일부 `None` 상태가 `NaN`으로 해석되는 것을 확인하였다.
향후 Annotation Guideline(라벨링 가이드라인)에서는 비적용 속성값을 
`None`으로 명시적으로 기록하여 실제 Missing Value(결측값)와 구분할 필요가 있다. 

**Overall Result: PASS**

## QA-04. Bounding Box Validation(바운딩 박스 검증)

### 목적

각 Object Labeling(객체 라벨링)에 생성된 Bounding Box(바운딩 박스)의 
좌표와 크기가 기하학적으로 유효한지 검증한다. 

잘못된 좌표 순서, 0 이하의 Width/Height(너비/높이), 음수 좌표 및 이미지 경계를
벗어난 Bounding Box(바운딩 박스)를 탐지하여 객체 위치 정보의 품질을 평가한다. 

### 주요 검증 항목 

- `x_min < x_max` 조건 확인
- `y_min < y_max` 조건 확인
- Bounding Box Width(바운딩 박스 너비) > 0
- Bounding Box Height(바운딩 박스 높이) > 0
- Negative Coordinate(음수 좌표) 여부 확인 
- Image Boundary(이미지 경계) 초과 여부 확인
- 비정상적인 Bounding Box Size(바운딩 박스 크기) 탐지

In [13]:
# Bounding Box Geometry(바운딩 박스 기하 구조) 검증

# Width(너비)와 Height(높이) 계산
df["bbox_width"] = df["x_max"] - df["x_min"]
df["bbox_height"] = df["y_max"] - df["y_min"]

# 잘못된 좌표 순서 및 크기 탐지
invalid_x = df["x_min"] >= df["x_max"]
invalid_y = df["y_min"] >= df["y_max"]

invalid_width = df["bbox_width"] <= 0
invalid_height = df["bbox_height"] <= 0

# 음수 좌표 탐지
negative_coordinates = (
    (df["x_min"] < 0) |
    (df["y_min"] < 0) |
    (df["x_max"] < 0) |
    (df["y_max"] < 0)
)

print("=== Bounding Box Geometry Validation ===")
print(f"Total Bounding Boxes: {len(df)}")
print(f"Invalid X Coordinates: {invalid_x.sum()}")
print(f"Invalid Y Coordinates: {invalid_y.sum()}")
print(f"Invalid Width: {invalid_width.sum()}")
print(f"Invalid Height: {invalid_height.sum()}")
print(f"Negative Coordinates: {negative_coordinates.sum()}")

geometry_issues = (
    invalid_x |
    invalid_y |
    invalid_width |
    invalid_height |
    negative_coordinates
)

print(f"\nBounding Boxes with Geometry Issues: {geometry_issues.sum()}")

if geometry_issues.sum() == 0:
    print("Result: PASS - All bounding boxes have valid geometry.")
else:
    print("Result: REVIEW - Invalid bounding box geometry detected.")

=== Bounding Box Geometry Validation ===
Total Bounding Boxes: 73
Invalid X Coordinates: 0
Invalid Y Coordinates: 0
Invalid Width: 0
Invalid Height: 0
Negative Coordinates: 0

Bounding Boxes with Geometry Issues: 0
Result: PASS - All bounding boxes have valid geometry.


In [14]:
# Images(이미지) 폴더 구조 확인

images_dir = project_root / "images"

print("Images Directory:")
print(images_dir)

print("\nImage Files / Folders:")
for item in sorted(images_dir.iterdir()):
    print("-", item.name)

Images Directory:
c:\Users\user\OneDrive\Documents\data-labeling-portfolio\03_autonomous_driving\images

Image Files / Folders:


In [15]:
# 프로젝트 내부 Image File(이미지 파일) 검색

image_extensions = {".jpg", ".jpeg", ".png"}

image_files = [
    path for path in project_root.rglob("*")
    if path.is_file() and path.suffix.lower() in image_extensions
]

print("=== Image File Search ===")
print(f"Total Image Files Found: {len(image_files)}")

for path in image_files:
    print("-", path.relative_to(project_root))

=== Image File Search ===
Total Image Files Found: 658
- data\pilot\n006-2018-09-17-12-15-45-0400__CAM_FRONT__1537201092012482.jpg
- data\pilot\n008-2018-03-14-15-16-29-0400__CAM_FRONT__1521055565282267.jpg
- data\pilot\n008-2018-05-21-11-06-59-0400__CAM_FRONT__1526915374762465.jpg
- data\pilot\n008-2018-09-18-14-18-33-0400__CAM_FRONT__1537294833612404.jpg
- data\pilot\n009-2018-09-12-09-59-51-0400__CAM_FRONT__1536761961012656.jpg
- data\pilot\n013-2018-08-01-16-46-39+0800__CAM_FRONT__1533113758512407.jpg
- data\pilot\n013-2018-08-20-14-38-24+0800__CAM_FRONT__1534747862612407.jpg
- data\pilot\n016-2018-07-04-10-44-39+0800__CAM_FRONT__1530672741162515.jpg
- data\raw\samples\CAM_BACK\n003-2018-01-03-12-03-23+0800__CAM_BACK__1514952316316487.jpg
- data\raw\samples\CAM_BACK\n003-2018-01-08-11-30-34+0800__CAM_BACK__1515382745757583.jpg
- data\raw\samples\CAM_BACK\n003-2018-07-12-15-40-35+0800__CAM_BACK__1531383596387543.jpg
- data\raw\samples\CAM_BACK\n005-2018-06-14-20-11-03+0800__CAM_BACK

### QA-04-3. Annotation–Image Matching Validation (라벨링–이미지 매칭 검증)

라벨링 데이터셋에 기록된 모든 이미지가 실제 Pilot Image Directory
(파일럿 이미지 디렉터리)에 존재하는지 검증한다.

이 검증을 통해 각 Bounding Box(바운딩 박스) 라벨링이 해당 원본 이미지와
정확하게 연결되어 있는지 확인하고, 이후 Image Boundary Validation
(이미지 경계 검증)을 수행하기 위한 데이터 정합성을 확보한다.

**검증 항목**
- 라벨링 데이터에 포함된 고유 이미지 수 확인
- Pilot Image Directory(파일럿 이미지 디렉터리)의 실제 이미지 수 확인
- 라벨링 데이터와 실제 이미지 파일 간 매칭 여부 확인
- 누락된 이미지 및 라벨링에서 참조되지 않은 이미지 확인

In [16]:
# Pilot Image Directory(파일럿 이미지 디렉터리)
pilot_dir = project_root / "data" / "pilot"

# CSV에 기록된 고유 Image Name(이미지명)
annotated_images = set(df["image_name"].unique())

# 실제 파일럿 이미지
pilot_images = {
    path.name
    for path in pilot_dir.glob("*")
    if path.suffix.lower() in {".jpg", ".jpeg", ".png"}
}

# Matching Validation(매칭 검증)
matched_images = annotated_images & pilot_images
missing_images = annotated_images - pilot_images
unreferenced_images = pilot_images - annotated_images

print("=== Annotation-Image Matching Validation ===")
print(f"Annotated Images: {len(annotated_images)}")
print(f"Pilot Image Files: {len(pilot_images)}")
print(f"Matched Images: {len(matched_images)}")
print(f"Missing Images: {len(missing_images)}")
print(f"Unreferenced Pilot Images: {len(unreferenced_images)}")

if missing_images:
    print("\nMissing Image Files:")
    for name in sorted(missing_images):
        print("-", name)

if unreferenced_images:
    print("\nUnreferenced Pilot Images:")
    for name in sorted(unreferenced_images):
        print("-", name)

if len(missing_images) == 0:
    print("\nResult: PASS - All annotated images were found.")
else:
    print("\nResult: FAIL - Some annotated images could not be found.")

=== Annotation-Image Matching Validation ===
Annotated Images: 8
Pilot Image Files: 8
Matched Images: 0
Missing Images: 8
Unreferenced Pilot Images: 8

Missing Image Files:
- image_01.jpg
- image_02.jpg
- image_03.jpg
- image_04.jpg
- image_05.jpg
- image_06.jpg
- image_07.jpg
- image_08.jpg

Unreferenced Pilot Images:
- n006-2018-09-17-12-15-45-0400__CAM_FRONT__1537201092012482.jpg
- n008-2018-03-14-15-16-29-0400__CAM_FRONT__1521055565282267.jpg
- n008-2018-05-21-11-06-59-0400__CAM_FRONT__1526915374762465.jpg
- n008-2018-09-18-14-18-33-0400__CAM_FRONT__1537294833612404.jpg
- n009-2018-09-12-09-59-51-0400__CAM_FRONT__1536761961012656.jpg
- n013-2018-08-01-16-46-39+0800__CAM_FRONT__1533113758512407.jpg
- n013-2018-08-20-14-38-24+0800__CAM_FRONT__1534747862612407.jpg
- n016-2018-07-04-10-44-39+0800__CAM_FRONT__1530672741162515.jpg

Result: FAIL - Some annotated images could not be found.


In [17]:
# Annotation–Image Mapping(라벨링–이미지 매핑) 정보 검색

mapping_keywords = [
    "image_01",
    "image_02",
    "image_03",
    "image_04",
    "image_05",
    "image_06",
    "image_07",
    "image_08"
]

search_extensions = {".csv", ".txt", ".md"}

mapping_candidates = []

for path in project_root.rglob("*"):
    if path.is_file() and path.suffix.lower() in search_extensions:
        try:
            text = path.read_text(encoding="utf-8", errors="ignore")

            if any(keyword in text for keyword in mapping_keywords):
                mapping_candidates.append(path)

        except Exception:
            pass

print("=== Annotation–Image Mapping Search ===")
print(f"Candidate Files Found: {len(mapping_candidates)}")

for path in mapping_candidates:
    print("-", path.relative_to(project_root))

=== Annotation–Image Mapping Search ===
Candidate Files Found: 1
- annotations\pilot\manual_annotations_73.csv


### QA-04-4. Image Mapping Recovery(이미지 매핑 복원)

Annotation-Image Matching Validation(라벨링-이미지 매칭 검증) 과정에서 
라벨링 데이터의 Alias Filename(별칭 파일명)과 실제 원본 이미지 파일명이 
직접 연결되지 않는 문제를 확인하였다. 

라벨링 데이터에서는 `image_01.jpg` ~ `image_08.jpg` 형식의 별칭을 사용하지만, 
Pilot Image Directory(파일럿 이미지 디렉터리)에는 nuScenes 원본 파일명이 
유지되어 있어 파일명으로 직접 매칭할 수 없다.

따라서 기존 Bounding Box(바운딩 박스)를 후보 이미지에 시각적으로 표시하여
각 별칭과 원본 이미지의 대응 관계를 확인하고 Mapping Table(매핑 테이블)을 
구축한다.

### QA Finding(품질검수 발견사항)

- 라벨링 데이터와 원본 이미지 간 Filename(파일명) 체계가 일치하지 않음
- 기존 Mapping Table(매핑 테이블)이 존재하지 않음
- 원본 데이터에 대한 Data Traceability(데이터 추적성) 개선이 필요함

In [18]:
# Pilot Image(파일럿 이미지) 목록 확인

from PIL import Image
import matplotlib.pyplot as plt

pilot_image_files = sorted([
    path for path in pilot_dir.iterdir()
    if path.suffix.lower() in {".jpg", ".jpeg", ".png"}
])

print("=== Pilot Images ===")

for i, path in enumerate(pilot_image_files, start=1):
    with Image.open(path) as img:
        print(
            f"Candidate {i}: "
            f"{path.name} | "
            f"Size: {img.width} x {img.height}"
        )

=== Pilot Images ===
Candidate 1: n006-2018-09-17-12-15-45-0400__CAM_FRONT__1537201092012482.jpg | Size: 1600 x 900
Candidate 2: n008-2018-03-14-15-16-29-0400__CAM_FRONT__1521055565282267.jpg | Size: 1600 x 900
Candidate 3: n008-2018-05-21-11-06-59-0400__CAM_FRONT__1526915374762465.jpg | Size: 1600 x 900
Candidate 4: n008-2018-09-18-14-18-33-0400__CAM_FRONT__1537294833612404.jpg | Size: 1600 x 900
Candidate 5: n009-2018-09-12-09-59-51-0400__CAM_FRONT__1536761961012656.jpg | Size: 1600 x 900
Candidate 6: n013-2018-08-01-16-46-39+0800__CAM_FRONT__1533113758512407.jpg | Size: 1600 x 900
Candidate 7: n013-2018-08-20-14-38-24+0800__CAM_FRONT__1534747862612407.jpg | Size: 1600 x 900
Candidate 8: n016-2018-07-04-10-44-39+0800__CAM_FRONT__1530672741162515.jpg | Size: 1600 x 900


### QA-04-5. Image Boundary Validation(이미지 경계 검증)

Pilot Image(파일럿 이미지)의 실제 Image Dimension(이미지 크기)을 확인한 결과,
8개 이미지 모두 동일한 **1600 × 900 pixels** 크기로 구성되어 있음을 확인하였다.

따라서 각 Bounding Box(바운딩 박스)의 좌표가 이미지의 유효 범위 내에 존재하는지 검증한다. 

### 검증 기준

- `x_min >= 0`
- `y_min >= 0`
- `x_max <= 1600`
- `y_max <= 900`
- Bounding Box(바운딩 박스)의 모든 좌표가 이미지 경계 내에 존재해야 함

In [19]:
# Image Boundary(이미지 경계) 검증

IMAGE_WIDTH = 1600
IMAGE_HEIGHT = 900

left_boundary_issue = df["x_min"] < 0
top_boundary_issue = df["y_min"] < 0
right_boundary_issue = df["x_max"] > IMAGE_WIDTH
bottom_boundary_issue = df["y_max"] > IMAGE_HEIGHT

boundary_issues = (
    left_boundary_issue |
    top_boundary_issue |
    right_boundary_issue |
    bottom_boundary_issue
)

print("=== Image Boundary Validation ===")
print(f"Image Size: {IMAGE_WIDTH} x {IMAGE_HEIGHT}")
print(f"Total Bounding Boxes: {len(df)}")

print(f"\nLeft Boundary Issues: {left_boundary_issue.sum()}")
print(f"Top Boundary Issues: {top_boundary_issue.sum()}")
print(f"Right Boundary Issues: {right_boundary_issue.sum()}")
print(f"Bottom Boundary Issues: {bottom_boundary_issue.sum()}")

print(f"\nBounding Boxes Outside Image Boundary: {boundary_issues.sum()}")

if boundary_issues.sum() == 0:
    print("Result: PASS - All bounding boxes are within image boundaries.")
else:
    print("Result: REVIEW - Bounding boxes outside image boundaries detected.")

=== Image Boundary Validation ===
Image Size: 1600 x 900
Total Bounding Boxes: 73

Left Boundary Issues: 0
Top Boundary Issues: 0
Right Boundary Issues: 0
Bottom Boundary Issues: 0

Bounding Boxes Outside Image Boundary: 0
Result: PASS - All bounding boxes are within image boundaries.


### QA-04-6. Bounding Box Size Validation(바운딩 박스 크기 검증)

Bounding Box(바운딩 박스)의 Width(너비), Height(높이), Area(면적)을 계산하여 
박스 크기의 분포를 확인하고 비정상적으로 작거나 큰 객체가 존재하는지 검토한다. 

작은 Bounding Box(바운딩 박스)가 반드시 라벨링 오류를 의미하는 것이 아니며, 
원거리 객체 또는 Small/Distant(소형·원거리) 객체일 가능성이 있으므로 
크기 이상치는 자동 오류로 판정하지 않고 Review Candidate(검토 후보)로 분류한다. 

### 주요 검증 항목

- Bounding Box Width(바운딩 박스 너비) 분포
- Bounding Box Height(바운딩 박스 높이) 분포
- Bounding Boa Area(바운딩 박스 면적) 분포
- 이미지 전체 면적 대비 Bounding Box Area Ratio(바운딩 박스 면적 비율)
- 비정상적인 크기의 Review Candidate(검토 후보) 식별

In [20]:
# Bounding Box Size(바운딩 박스 크기) 계산

df["bbox_area"] = df["bbox_width"] * df["bbox_height"]

image_area = IMAGE_WIDTH * IMAGE_HEIGHT
df["bbox_area_ratio"] = df["bbox_area"] / image_area

print("=== Bounding Box Size Summary ===")

size_summary = df[
    ["bbox_width", "bbox_height", "bbox_area", "bbox_area_ratio"]
].describe()

display(size_summary)

print("\n=== Extreme Values ===")
print(f"Minimum Width: {df['bbox_width'].min():.2f}")
print(f"Minimum Height: {df['bbox_height'].min():.2f}")
print(f"Minimum Area: {df['bbox_area'].min():.2f}")

print(f"\nMaximum Width: {df['bbox_width'].max():.2f}")
print(f"Maximum Height: {df['bbox_height'].max():.2f}")
print(f"Maximum Area: {df['bbox_area'].max():.2f}")

print(
    f"\nMaximum Image Area Ratio: "
    f"{df['bbox_area_ratio'].max() * 100:.2f}%"
)

=== Bounding Box Size Summary ===


,bbox_width,bbox_height,bbox_area,bbox_area_ratio
count,73.000000,73.000000,73.000000,73.000000
mean,73.181096,66.287671,8756.893036,0.006081
std,80.994531,64.464005,21260.274821,0.014764
min,5.010000,12.720000,81.562800,0.000057
25%,16.860000,27.170000,668.283000,0.000464
50%,41.400000,44.260000,1682.971500,0.001169
75%,91.600000,76.940000,6910.488000,0.004799
max,421.850000,365.930000,154367.570500,0.107200



=== Extreme Values ===
Minimum Width: 5.01
Minimum Height: 12.72
Minimum Area: 81.56

Maximum Width: 421.85
Maximum Height: 365.93
Maximum Area: 154367.57

Maximum Image Area Ratio: 10.72%


### QA-04-7. Small/Distant–Bounding Box Size Consistency
(소형·원거리 속성–바운딩 박스 크기 일관성 검증)

Bounding Box Size Validation(바운딩 박스 크기 검증) 결과를
`small_distant` 속성과 비교하여 라벨링 기준의 일관성을 확인한다.

Small/Distant(소형·원거리)로 라벨링된 객체는 일반 객체보다
Bounding Box(바운딩 박스)의 크기가 상대적으로 작을 것으로 예상된다.

단, Bounding Box Size(바운딩 박스 크기)만으로 객체의 실제 거리나
라벨링 오류를 확정할 수 없으므로 본 검증에서는 두 그룹의 크기 분포를
비교하고 잠재적인 Review Candidate(검토 후보)를 식별한다.

### 주요 검증 항목

- `small_distant=True`와 `False`의 객체 수
- 두 그룹의 평균 및 중앙 Bounding Box Area(바운딩 박스 면적)
- 두 그룹의 Bounding Box Area Ratio(바운딩 박스 면적 비율)
- 속성과 Bounding Box Size(바운딩 박스 크기)의 전반적인 일관성

In [21]:
# Small/Distant(소형·원거리)와 Bounding Box Size(바운딩 박스 크기) 비교

size_by_distance = (
    df.groupby("small_distant")
    .agg(
        object_count=("object_id", "count"),
        mean_width=("bbox_width", "mean"),
        median_width=("bbox_width", "median"),
        mean_height=("bbox_height", "mean"),
        median_height=("bbox_height", "median"),
        mean_area=("bbox_area", "mean"),
        median_area=("bbox_area", "median"),
        mean_area_ratio=("bbox_area_ratio", "mean")
    )
    .round(2)
)

print("=== Small/Distant vs Bounding Box Size ===")
display(size_by_distance)

=== Small/Distant vs Bounding Box Size ===


,object_count,mean_width,median_width,mean_height,median_height,mean_area,median_area,mean_area_ratio
small_distant,,,,,,,,
False,32,122.85,92.7,107.31,90.03,18269.57,8090.12,0.01
True,41,34.41,31.8,34.27,28.09,1332.37,1025.39,0.00


### QA-04-8. Size Consistency Review Candidate
(크기 일관성 검토 후보)

Small/Distant(소형·원거리) 속성과 Bounding Box Size(바운딩 박스 크기)의
전체적인 관계는 일관된 것으로 확인되었다.

추가적으로 두 그룹의 Median Area(중앙 면적)를 기준으로 반대 그룹의
일반적인 크기 특성을 보이는 객체를 Review Candidate(검토 후보)로 식별한다.

### Review 기준

- `small_distant=True`이면서 Bounding Box Area(바운딩 박스 면적)가
  `False` 그룹의 Median Area(중앙 면적) 이상인 객체
- `small_distant=False`이면서 Bounding Box Area(바운딩 박스 면적)가
  `True` 그룹의 Median Area(중앙 면적) 이하인 객체

본 기준은 Annotation Error(라벨링 오류)를 자동 판정하기 위한 것이 아니라,
속성 라벨과 객체 크기의 관계가 예외적으로 나타나는 사례를 선별하여
Visual QA(시각적 품질검수) 대상으로 활용하기 위한 것이다.

In [22]:
# 그룹별 Median Area(중앙 면적)
median_area_false = df.loc[
    df["small_distant"] == False, "bbox_area"
].median()

median_area_true = df.loc[
    df["small_distant"] == True, "bbox_area"
].median()

# Review Candidate(검토 후보) 추출
large_but_small_distant = df[
    (df["small_distant"] == True) &
    (df["bbox_area"] >= median_area_false)
].copy()

small_but_not_distant = df[
    (df["small_distant"] == False) &
    (df["bbox_area"] <= median_area_true)
].copy()

print("=== Size Consistency Review Candidates ===")
print(f"False Group Median Area: {median_area_false:.2f}")
print(f"True Group Median Area: {median_area_true:.2f}")

print(
    "\nTrue but >= False Median Area: "
    f"{len(large_but_small_distant)}"
)

print(
    "False but <= True Median Area: "
    f"{len(small_but_not_distant)}"
)

print(
    "\nTotal Review Candidates: "
    f"{len(large_but_small_distant) + len(small_but_not_distant)}"
)

=== Size Consistency Review Candidates ===
False Group Median Area: 8090.12
True Group Median Area: 1025.39

True but >= False Median Area: 0
False but <= True Median Area: 3

Total Review Candidates: 3


In [23]:
# Review Candidate(검토 후보) 상세 확인

review_candidates = pd.concat([
    large_but_small_distant,
    small_but_not_distant
]).copy()

review_columns = [
    "image_name",
    "object_id",
    "class_label",
    "bbox_width",
    "bbox_height",
    "bbox_area",
    "small_distant",
    "occlusion",
    "truncation",
    "ambiguity"
]

review_candidates = (
    review_candidates[review_columns]
    .sort_values("bbox_area")
)

print("=== Size Consistency Review Candidate Details ===")
print(f"Total Candidates: {len(review_candidates)}")

display(review_candidates)

=== Size Consistency Review Candidate Details ===
Total Candidates: 3


,image_name,object_id,class_label,bbox_width,bbox_height,bbox_area,small_distant,occlusion,truncation,ambiguity
29,image_04.jpg,C04_OBJ_003,Traffic Cone,6.14,21.38,131.2732,False,NaN,NaN,NaN
69,image_08.jpg,C08_OBJ_005,Traffic Cone,9.40,17.28,162.4320,False,Partial,NaN,Boundary
30,image_04.jpg,C04_OBJ_004,Traffic Cone,14.16,27.17,384.7272,False,NaN,NaN,NaN


### QA-04 Size Validation Finding(크기 검증 결과)

Bounding Box Size(바운딩 박스 크기)와 Small/Distant(소형·원거리) 속성을
교차 검증한 결과, 두 그룹 간 뚜렷한 크기 차이가 확인되었다.

- `small_distant=False`: Median Area(중앙 면적) **8,090.12 px²**
- `small_distant=True`: Median Area(중앙 면적) **1,025.39 px²**
- `small_distant=True`이면서 일반 객체 수준으로 큰 객체: **0건**
- `small_distant=False`이면서 소형·원거리 그룹 수준으로 작은 객체: **3건**

3개의 Review Candidate(검토 후보)는 모두 Traffic Cone(교통콘)으로 확인되었다.
Traffic Cone은 객체 자체의 물리적 크기가 작기 때문에 Bounding Box Size
(바운딩 박스 크기)만으로 `small_distant` 속성 오류를 판단하기 어렵다.

특히 `C08_OBJ_005`는 Partial Occlusion(부분 가림)과 Boundary Ambiguity
(경계 모호성)가 함께 기록되어 있어 후속 Visual QA(시각적 품질검수)의
우선 검토 대상으로 지정한다.

**Result: REVIEW - 3 candidates identified for Visual QA; no automatic annotation error confirmed.**